lib/kube/proxy/sess.go

```diff
@@ -0,0 +1,1193 @@
+		PresenceEnabled:   ctx.Identity.GetIdentity().MFAVerified != "",
+		stateUpdate:       sync.NewCond(&sync.Mutex{}),
+	}
+
+	err = s.trackerCreate(initiator, policySets)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return s, nil
+}
+
+// waitOnAccess puts the session in pending mode and waits for the session
+// to fulfill the access requirements again.
+func (s *session) waitOnAccess() {
+	s.io.Off()
+	s.io.BroadcastMessage("Session paused, Waiting for required participants...")
+
+	s.stateUpdate.L.Lock()
+	defer s.stateUpdate.L.Unlock()
+
+outer:
+	for {
+		switch s.state {
+		case types.SessionState_SessionStatePending:
+			continue
+		case types.SessionState_SessionStateTerminated:
+			return
+		case types.SessionState_SessionStateRunning:
+			break outer
+		}
+
+		s.stateUpdate.Wait()
+	}
+
+	s.io.BroadcastMessage("Resuming session...")
+	s.io.On()
+}
+
+// checkPresence checks the presence timestamp of involved moderators
+// and kicks them if they are not active.
+func (s *session) checkPresence() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	sess, err := s.trackerGet()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for _, participant := range sess.GetParticipants() {
+		if participant.ID == s.initiator.String() {
+			continue
+		}
+
+		if participant.Mode == string(types.SessionModeratorMode) && time.Now().UTC().After(participant.LastActive.Add(PresenceMaxDifference)) {
+			s.log.Debugf("Participant %v is not active, kicking.", participant.ID)
+			id, _ := uuid.Parse(participant.ID)
+			err := s.leave(id)
+			if err != nil {
+				s.log.WithError(err).Warnf("Failed to kick participant %v for inactivity.", participant.ID)
+			}
+		}
+	}
+
+	return nil
+}
+
+// launch waits until the session meets access requirements and then transitions the session
+// to a running state.
+func (s *session) launch() error {
+	defer func() {
+		err := s.Close()
+		if err != nil {
+			s.log.WithError(err).Errorf("Failed to close session: %v", s.id)
+		}
+	}()
+
+	go func() {
+		select {
+		case <-time.After(time.Until(s.expires)):
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			s.io.BroadcastMessage("Session expired, closing...")
+
+			err := s.Close()
+			if err != nil {
+				s.log.WithError(err).Error("Failed to close session")
+			}
+		case <-s.closeC:
+		}
+	}()
+
+	// If the identity is verified with an MFA device, we enabled MFA-based presence for the session.
+	if s.PresenceEnabled {
+		go func() {
+			ticker := time.NewTicker(PresenceVerifyInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ticker.C:
+					err := s.checkPresence()
+					if err != nil {
+						s.log.WithError(err).Error("Failed to check presence, closing session as a security measure")
+						err := s.Close()
+						if err != nil {
+							s.log.WithError(err).Error("Failed to close session")
+						}
+					}
+				case <-s.closeC:
+					return
+				}
+			}
+		}()
+	}
+
+	s.log.Debugf("Launching session: %v", s.id)
+	s.mu.Lock()
```

Comment:
Looks like we may be holding on to this mutex for a long time (including while we're making network calls to emit audit events). We okay with this?

Reply:
Reduced the lock time significantly now